text/1042-fastapi.rst

```diff
@@ -0,0 +1,357 @@
+
+       .. code-block:: python
+
+           class User:
+
+               name: str
+               friends: list[User__friends__User]
+
+    b. where ``User__friends__User`` will be a special variant of ``User`` reachable by traversing the ``User.friends`` link:
+
+       .. code-block:: python
+
+           class User__friends__User(User):
+
+               class __linkprops__:
+                   weight: int | None
+
+  - This enables link properties to always be fetched and never shadow regular properties or links. Accessing them will be possible via the ``__linkprops__`` property, e.g.:
+
+    .. code-block:: python
+
+        user = await db.get(User.select(friends=User.friends.select()))
+
+        weight = user.friends.__linkprops__.weight  # type: `int | None`
+
+* ``.id`` is a required property in Gel, however, in Python *new* objects won't have an ``.id`` until they pass through a ``save()`` call. This presents a dilemma: should the ORM type's ``id`` property be optional or required? We say it will be required, because objects will be read more often than written, so we want to optimize for the common case. Accessing the ``id`` property on an unsaved object will raise a runtime error.
+
+  - That said, the reflected types will have a custom ``__init__`` that will not accept an ``id`` argument (in disagreement with the type annotation for the field being "required").
+
+  - The custom ``__init__`` will also require passing all required links, which would be in disagreement with all links being "optional" in Python by default.
+
+* Prisma-style API for custom shapes: ``User.select(name=True)``, ``User.select(name=True, settings=User.settings.select())`` instead of ``User.select(User.name)``.
+
+* We will prohibit shadowing links with detached types, e.g. this is correct:
+
+  .. code-block:: python
+
+      User.select(name=True, friends=User.friends.select())
+
+  and this is incorrect and will raise an error:
+
+  .. code-block:: python
+
+      User.select(name=True, friends=User.select())
+
+  (only in Python ORM, EdgeQL will still allow this.)
+
+* Accessing a property or link that wasn't fetched will raise an error.
+
+  - But, setting a property or link that wasn't fetched will *not* raise an error:
+
+    .. code-block:: python
+
+        user = await db.get(User.select(name=True))
+
+        print(user.email)  # <- runtime error
+
+        user.email = 'peter@example.com'  # <- fine!
+        await db.save(user)
+
+* "Multi" links and properties will be represented as list-like collections in Python:
+
+  - Properties will be represented as ``gel.List`` (the exact name is still TBD) and will allow duplicate entries.
+
+  - Links will be represented as ``gel.DistinctList`` and will not allow duplicate entries.
+
+  - ``List`` and ``LinkList``, like Python's ``list`` will implement the ``+=`` operator to extend the list, e.g. ``user.friends += [friend]`` (note the square brackets on the right hand side).
```

Comment:
I assume that here and below `LinkList` is meant to be `DistinctList`. In any case, I would vote strongly against `LinkList` as a name, since it's so similar to a "linked list", which makes it unnecessarily confusing.

Reply:
Hm, i started updating it and then stumbled upon this (which prompted me to use LinkList in the first place):

>   - ``LinkList`` will have a custom-tailored ``append()`` method that would accept keyword-only arguments to set link properties, e.g. ``user.friends.append(friend, weight=10)``.

So `DistinctList` wouldn't work. It should be either `LinkList` or we coin a better name. The silver lining here is that the name of the class is kinda internal here, users shouldn't care or see it ever.